pkg/cache/tas_flavor_snapshot.go

```diff
@@ -244,6 +245,67 @@ func (s *TASFlavorSnapshot) removeTASUsage(domainID utiltas.TopologyDomainID, us
 	s.leaves[domainID].tasUsage.Sub(usage)
 }
 
+func (s *TASFlavorSnapshot) freeCapacityPerDomain() map[utiltas.TopologyDomainID]resources.Requests {
+	freeCapacityPerDomain := make(map[utiltas.TopologyDomainID]resources.Requests, len(s.leaves))
+
+	for domainID, leaf := range s.leaves {
+		freeCapacityPerDomain[domainID] = leaf.freeCapacity.Clone()
+	}
+
+	return freeCapacityPerDomain
+}
+
+func (s *TASFlavorSnapshot) tasUsagePerDomain() map[utiltas.TopologyDomainID]resources.Requests {
+	tasUsagePerDomain := make(map[utiltas.TopologyDomainID]resources.Requests, len(s.leaves))
+
+	for domainID, leaf := range s.leaves {
+		tasUsagePerDomain[domainID] = leaf.tasUsage.Clone()
+	}
+
+	return tasUsagePerDomain
+}
+
+type domainCapacityDetails struct {
+	FreeCapacity map[corev1.ResourceName]string `json:"freeCapacity"`
+	TasUsage     map[corev1.ResourceName]string `json:"tasUsage"`
+}
+
+func (s *TASFlavorSnapshot) SerializeFreeCapacityPerDomain() (string, error) {
```

Comment:
I think this is quite useful for logging, as part of the https://github.com/kubernetes-sigs/kueue/issues/4091 (second point).

Reply:
cc @PBundyra